### ISSUE TITLE
Make it cheaper to spawn isolates with overlapping code to the current isolate

### ISSUE THREAD

**Issue by natebosch** [dart-lang/sdk#32953]
If we are spawning isolate which will include imports to the same libraries that are already loaded in the VM it would be nice to not pay the cost all over again to parse the Dart source.

This will be especially useful for `package:test` where we spawn an Isolate for each VM test we run.

cc @mraleph

**Comment by mraleph** [comment]
Currently when creating new isolates we transparently parse Dart source code into Kernel using DFE isolate - without sharing or preserving any state between compilation requests.

@jensjoha work on Flutter test setup shows that significant performance improvement can be achieved by preserving the state between compilation requests.

I wonder if we can make it work for `package:test` too - maybe we need to extend `spawnUri` API.

/cc @lrhn

**Comment by lrhn** [comment]
The way `spawn` and `spawnUri` have so far been defined (but not necessarily implemented), the `spawn` function must run exactly the same code as the spawning isolate. It's an error to re-load/fetch any source if that allows it to change.

For `spawnUri`, there is no such requirement, and fetching the source again is allowed, but not required. You can do all the caching you would want. It's somewhat rare that the entry-point library already exists in the spawning isolate (then you could probably just have used `spawn`), but not so unlikely to share utility libraries.

I don't see what changes the `spawnUri` method would be beneficial here.

**Comment by mraleph** [comment]
@lrhn currently VM completely discards CFE state between URI loading requests because it has no way of knowing whether there will ever be another URI loading request that might benefit from the cached state - and discarding the state frees significant chunk of memory. So the change we could do to `spawnUri` is to add a way to hint to the VM that it should not discard the state - because another request is likely. This is one approach. 

Another approach is to change `package:test` to explicitly depend on the CFE and explicitly manage its compilation requests just like `flutter test` does.

**Comment by matanlurey** [comment]
My understanding is we want to _remove_ having to understand compilation pipelines from `package:test`, not further complicate it :) ... I think @natebosch has considered supporting the CFE as part of the build system, WDUT?

**Comment by natebosch** [comment]
we might not be able to get away from having an understanding of the CFE from within package:test

Running VM tests is explicitly depending on running from source. Paying the parsing costs for _every_ test file is going to be a terrible experience for our users. We'd either need to make everyone go through `build_runner` to run tests, or teach the test package about the CFE.

Based on my understanding of what flutter is doing I don't think it would be terribly costly to implement. The cost comes in when we think about version constraints. Flutter is pinned to an SDK, package:test is not.

**Comment by dgrove** [comment]
/cc @a-siva 

@a-siva @mraleph who's the right person to own this?

**Comment by mraleph** [comment]
@dgrove 

First we need consensus how to implement it: 

1. teach `package:test` to use `package:front_end` like Flutter does. might require expanding Isolate spawn APIs to spawn Isolates from dill files. Should probably be jointly owned by CFE team and Nate.
2. expand `spawnUri` API to communicate to the DFE service that it needs to preserve the state of CFE in memory. Should probably be owned by CFE/VM team and Lasse.

I think solution number 1 is preferable because extending `spawnUri` in a flexible and future proof way is hard.

**Comment by dgrove** [comment]
Self-assigning until we get consensus on what to implement.

**Comment by grouma** [comment]
I'm not familiar with how Flutter uses `package:front_end`. @mraleph can you easily expand on option 1? If not we can set some time up over GVC to discuss it in detail.

**Comment by natebosch** [comment]
If someone can share a link to the flutter test code using this pattern we can build a prototype in package:test and evaluate it's feasibility.

**Comment by mraleph** [comment]
@natebosch @grouma the logic is in [this file](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/flutter_platform.dart).

Each test is compiled using persistent compiler [here](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/flutter_platform.dart#L334-L335) and then run using `flutter_tester` process [here](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/test/flutter_platform.dart#L344-L353).

Let me know if you need any help. Another person to chat with might be @aam.

**Comment by aam** [comment]
Yes, happy to help, but am picking something from the earlier responses where @lrhn wrote:
> The way spawn and spawnUri have so far been defined (but not necessarily implemented), the spawn function must run exactly the same code as the spawning isolate. It's an error to re-load/fetch any source if that allows it to change.

Not sure what scenario last statement in the quote refers to. For `spawn` are we prohibiting hot-reload for spawned and/or spawning isolates?

**Comment by lrhn** [comment]
I was not referring to "hot reload", merely saying that `Isolate.spawn` should effectively run exactly the same source in both spawning and spawned isolate. The spawn operation should not fetch any source code from a file or a web server, since it isn't guaranteed to get the same source again.

What you do with hot-reload during development is not covered by the language or library specifications, it's just a best-effort attempt to avoid restarting the program from scratch.

**Comment by grouma** [comment]
Tried to take a look at this today. We need to wait for the next SDK release as we are blocked by https://github.com/dart-lang/sdk/issues/32950.

**Comment by dgrove** [comment]
#32950 is fixed now, @grouma . Are you able to take another look?

**Comment by dgrove** [comment]
@grouma ping.

**Comment by natebosch** [comment]
Direct integration with `front_end` from `test` is not an approach we are pursuing today. We're focusing on a modular compilation to kernel and we'll be testing the performance of a modular kernel approach for flutter test as well.

**Comment by dgrove** [comment]
Can this be moved from Dart2Stable ?

**Comment by natebosch** [comment]
This should not be a blocker for Dart 2 stable. I think this is still worth exploring a solution directly in the VM in a future milestone. Pushing complexity on every consumer of the `spawnUri` API in order to work around performance issues that could be solved once seems better in the long term than having many potential consumers of `package:front_end`.

**Comment by mraleph** [comment]
@natebosch only the caller of the `spawnUri` knows if the spawned Uri is going to be related in any way to other Uri-s that has been spawned before. What we need here is an ability to communicate to CFE instance within the VM that you are expecting to have an intersection between what it compiled before and what it is going to compile in the future. 

Unconditionally keeping CFE state around would double the memory overhead - so we can't just do that to sometimes make some subsequent `spawnUri`-s faster.

We could optionally extend some of the APIs to communicate the need to persist the CFE state (as I mentioned in my [comment above](https://github.com/dart-lang/sdk/issues/32953#issuecomment-385232218)). 

But I am not sure VM itself would benefit from such complexity. 

To be honest I would prefer we entirely remove `spawnUri` that allows one to spawn isolates from raw Dart sources.

**Comment by joeconwaystk** [comment]
Some additional usage info on `spawnUri`: we use it often in CLI tools to generate and run scripts within the context of a dependent application. For example, `aqueduct serve` generates a script that imports the Dart application package in current working directory and starts the application defined by a type located in that package. 

If it aids the conversation, [this package](https://github.com/stablekernel/dart-isolate-executor) is a minimal implementation of this technique (and is what is used by these CLI tools).
